File: volume-26/2609/2609.rb
def calc(w, h, v, t, x, y, p, q)
  c = 0
  d = v * t
  (-d / w - 2 .. d / w + 2).each do |i|
    tx = i * w + (i.odd? ? w - p : p)
    next if (tx - x).abs > d
    dy = Math::sqrt(d ** 2 - (tx - x) ** 2);
    next if (q - y).abs > dy
    r = [ (dy + y) / h, -(dy - y) / h - 1 ].map(&:to_i)
    r[1] -= 1 if y > dy
    if r[0] == r[1]
      c += 1 if (tx - x) ** 2 + (q - y) ** 2 <= d * d
    else
      c += r[0] - r[1] - 1;
      r.each do |k|
        ty = k * h + (k.odd? ? h - q : q);
        c += 1 if (tx - x) ** 2 + (ty - y) ** 2 <= d * d
      end
    end
  end
  c
end
1.times do
  w, h, v, t, x, y, p, q = gets.split.map(&:to_i)
  puts calc(w, h, v, t, x, y, p, q)
end

File: volume-26/2609/2609.cc
#include <iostream>
#include <cmath>
using namespace std;

long long calc(int w, int h, int v, int t, int x, int y, int p, int q) {
  long long tx, ty, c = 0, d = v * t;
  for (int i = -d / w - 2; i < d / w + 2; ++i) {
    tx = i * w + (i % 2 ? w - p : p);
    if (abs(tx - x) > d) continue;
    double dy = sqrt(d * d - (tx - x) * (tx - x));
    if (abs(q - y) > dy) continue;

    int r[] = { (dy + y) / h, -(dy - y) / h - 1 };
    if (y > dy) r[1]--;
    if (r[0] == r[1]) {
      if ((tx - x) * (tx - x) + (q - y) * (q - y) <= d * d) c++;
    } else {
      c += r[0] - r[1] - 1;
      for (int k = 0; k < 2; ++k) {
        ty = r[k] * h + (r[k] % 2 ? h - q : q);
        if ((tx - x) * (tx - x) + (ty - y) * (ty - y) <= d * d) c++;
      }
    }
  }
  return c;
}

int main() {
  int w, h, v, t, x, y, p, q;
  cin >> w >> h >> v >> t >> x >> y >> p >> q;
  cout << calc(w, h, v, t, x, y, p, q) << endl;
}
